### PR TITLE
Add Storage trait for PbftState

### PR DESCRIPTION
**Pull request by knkski**
Introduces `Storage` trait that ensures that modifications to an object are synced to a backing store using RAII, and wraps `PbftState` in that. Also moves `PbftState` to be state managed by the engine, for two reasons:

 * `PbftNode` requiring `&mut self` on its methods and also having it hand out references to a contained `PbftState` didn't play nicely with lifetimes
 *  Syncing changes only at the end of each engine loop means we don't have to worry about whether or not any state changes within an engine loop need to be atomic, each loop either gets synced or not

### CODE REVIEW

File: src/storage.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use serde_json::{from_str, to_string};
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Write, Seek, SeekFrom};
+use std::ops::{Deref, DerefMut};
+use std::fmt;
+
+/// TODO: Write some docs
```
Comment (peterschwarz) [inline src/storage.rs]: TODO should be removed

File: build.rs [hunk 2]
```diff
@@ -36,7 +36,10 @@ fn main() {
         out_dir: &dest_path.to_str().unwrap(),
         input: &[proto_path.join("pbft_message.proto").to_str().unwrap()],
         includes: &[proto_path.to_str().unwrap()],
-        customize: Customize::default(),
+        customize: Customize {
```
Comment (peterschwarz) [inline build.rs]: This, and the `Cargo.toml` changes should be it's own commit.

File: src/storage.rs [hunk 3]
```diff
@@ -94,101 +121,226 @@ impl<T: Serialize + DeserializeOwned> Storage<T> {
...
+        Box::new(DiskStorageReadGuard::new(self))
+    }
+
+    fn write(&'a mut self) -> Box<StorageWriteGuard<'a, T, Target = T> + 'a> {
+        Box::new(DiskStorageWriteGuard::new(self))
+    }
+}
+
+// Mem Storage-related structs and impls
+/// TODO: Write some docs
```
Comment (peterschwarz) [inline src/storage.rs]: Another TODO to remove (or do)

File: src/engine.rs [hunk 4]
```diff
@@ -60,29 +62,33 @@ impl Engine for PbftEngine {
             .expect("This node is not in the peers list, which is necessary")
             as u64;
 
+        let mut pbft_state =
+            DiskStorage::new_with_default("test.json", || PbftState::new(node_id, &config))
```
Comment (bridger-herman) [inline src/engine.rs]: I realize this is WIP, but I still think a name more descriptive than `"test.json"` may be in order here

File: src/storage.rs [hunk 5]
```diff
@@ -94,101 +121,226 @@ impl<T: Serialize + DeserializeOwned> Storage<T> {
...
+#[derive(Debug)]
+pub struct MemStorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
+    __storage: &'a MemStorage<T>,
```
Comment (bridger-herman) [inline src/storage.rs]: What is the purpose of the leading underscores on this variable?

File: src/storage.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    }
+}
+
+impl<'a, T: 'a + Serialize + DeserializeOwned + fmt::Display> fmt::Display for StorageReadGuard<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+
+/// TODO: Write some docs
```
Comment (aludvik) [inline src/storage.rs]: Remove todo comment, create story instead

File: src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    }
+}
+
+impl<'a, T: 'a + Serialize + DeserializeOwned + fmt::Display> fmt::Display for StorageWriteGuard<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+
+/// TODO: Write some docs
```
Comment (aludvik) [inline src/storage.rs]: Remove todo comment, write story instead.

File: src/storage.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                    .map_err(|err| format!("Couldn't read file: {}", err))?
+            }
+            Err(_) => default()
+        };
+
+        // Then open the file again and truncate, preparing it to be written to
+        Ok(Self { data, file: OpenOptions::new()
+            .read(true)
+            .write(true)
+            .truncate(true)
```
Comment (aludvik) [inline src/storage.rs]: I do not think truncating the file immediately is a good idea. This would mean that if the process were to crash after reading the file into memory, but before a new value is written, the persisted data would be lost.

File: src/engine.rs (hunk 4 above)
Comment (aludvik) [reply src/engine.rs]: This is what we have done in the past:
https://github.com/hyperledger/sawtooth-raft/blob/master/src/path.rs
https://github.com/hyperledger/sawtooth-raft/blob/master/src/fs_storage.rs#L36

It may also be a good idea to split up the state across several files, so that we are not writing all of state to disk every time we change any part of it.

File: src/storage.rs [hunk 9]
```diff
@@ -1,85 +1,112 @@
...
+    fn write(&'a mut self) -> Box<StorageWriteGuard<'a, Self::S, Target = Self::S> + 'a>;
+}
+
+// Disk Storage-related structs and impls
 /// TODO: Write some docs
 #[derive(Debug)]
-pub struct StorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
-    cache: &'a Storage<T>,
+pub struct DiskStorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
+    __storage: &'a DiskStorage<T>,
```
Comment (aludvik) [inline src/storage.rs]: Why are you using two underscores in front of the variable name here? The standing convention is to let the compiler enforce public vs. private.

File: src/storage.rs [hunk 10]
```diff
@@ -1,85 +1,112 @@
...
-use std::io::{Read, Write, Seek, SeekFrom};
+use std::io::{Read, Seek, SeekFrom, Write};
 use std::ops::{Deref, DerefMut};
-use std::fmt;
 
+// General traits
+pub trait StorageReadGuard<'a, T>: Deref<Target = T> {}
+pub trait StorageWriteGuard<'a, T>: DerefMut<Target = T> {}
+
+pub trait Storage<'a>: fmt::Display {
```
Comment (aludvik) [inline src/storage.rs]: Why is this lifetime required? I understand that the lifetimes below are needed to guarantee that the guards live as long as the storage.

File: src/storage.rs (hunk 6 above)
Comment (knkski) [reply src/storage.rs]: Buried this at the end of the PR description, but I was planning on filling out the doc TODOs before this would get merged, I wanted to get feedback about the general approach first though

File: src/storage.rs (hunk 10 above)
Comment (aludvik) [reply src/storage.rs]: Does this work for you? It works locally for me, but I am just compiling storage.rs because the rest doesn't compile because of the protobuf/serde stuff in the sdk.

https://github.com/aludvik/sawtooth-pbft-1/tree/state-caching-no-lifetime

File: src/storage.rs (hunk 9 above)
Comment (knkski) [reply src/storage.rs]: Borrowed from https://doc.rust-lang.org/src/std/sync/rwlock.rs.html#121, but I can't seem to find a solid reason that they do it there, so I'll just switch back to no leading underscores

File: src/config.rs [hunk 11]
```diff
@@ -75,6 +79,7 @@ impl PbftConfig {
 /// + `sawtooth.consensus.pbft.view_change_timeout` (optional, default 4000 ms)
 /// + `sawtooth.consensus.pbft.message_timeout` (optional, default 100 blocks)
 /// + `sawtooth.consensus.pbft.max_log_size` (optional, default 1000 messages)
+/// + `sawtooth.consensus.pbft.storage` (optional, default `"memory"`)
```
Comment (aludvik) [inline src/config.rs]: I do not think storage type should be an on-chain setting.

File: src/node.rs [hunk 12]
```diff
@@ -44,34 +44,26 @@ pub struct PbftNode {
     /// Used for interactions with the validator
     pub service: Box<Service>,
 
-    /// Storage of state information
-    pub state: PbftState,
-
```
Comment (aludvik) [inline src/node.rs]: Refactoring the codebase to pass state around instead of passing it as an instance variable isn't within the scope of the commit message and should be split into a separate commit. I am also not certain this is a good change, so it would be good to see some reasoning behind it.

File: src/timing.rs [hunk 13]
```diff
@@ -52,10 +52,11 @@ enum TimeoutState {
 
 /// A timer that expires after a given duration
 /// Check back on this timer every so often to see if it's expired
-#[derive(Debug)]
+#[derive(Debug, Serialize, Deserialize)]
```
Comment (aludvik) [inline src/timing.rs]: Deriving serde support should be a separate commit

File: src/config.rs (hunk 11 above)
Comment (knkski) [reply src/config.rs]: I borrowed this from where it looked like Raft was going, in that it has storage configuration wrapped up in the `Config` class. Are you thinking of configuring this with a CLI or config file parameter? If config file, would it make sense to have the SDK handle that?

File: src/node.rs (hunk 12 above)
Comment (knkski) [reply src/node.rs]: Moving `PbftState` into the `Storage` wrapper is pretty tied up with moving it into state managed by the engine. Would it make sense to just expand the commit description?

File: src/config.rs (hunk 11 above)
Comment (aludvik) [reply src/config.rs]: I don't think the storage type should be a run-time configuration. I think the modularity is helpful to allow us to swap out different implementations later on more easily.

File: src/storage/disk.rs [hunk 14]
```diff
@@ -0,0 +1,149 @@
+//! Disk-backed persistence wrapper
```
Comment (agunde406) [inline src/storage/disk.rs]: Missing license

File: src/storage/memory.rs [hunk 15]
```diff
@@ -0,0 +1,112 @@
+//! Memory-backed persistence wrapper
```
Comment (agunde406) [inline src/storage/memory.rs]: Missing license

File: src/storage/mod.rs [hunk 16]
```diff
@@ -0,0 +1,216 @@
+//! Storage trait for syncing writes to an object to a backing store
```
Comment (agunde406) [inline src/storage/mod.rs]: Missing license
